File: pictionary-app/src/reducers/game.js
/* eslint-disable camelcase */
import { SAVE_GAME, UPDATE_GAME_STATE, UPDATE_GAME_PLAYERS, ADMIN_UPDATED, REMOVE_PLAYER } from '../constants/actionTypes';

const initialState = {
  id: null,
  rounds: 3,
  time: 80,
  max_players: 10,
  custom_words: '',
  custom_words_probability: 50,
  public_game: true,
  vote_kick_enabled: true,
  players: [],
  started: false,
  creator_id: null
};

const gameReducer = (state = initialState, action) => {
  switch (action.type) {
    case SAVE_GAME:
      const {
        id,
        rounds,
        time,
        max_players,
        custom_words,
        custom_words_probability,
        public_game,
        vote_kick_enabled,
        players,
        started,
        creator_id
      } = action.payload;
      return {
        ...state,
        id,
        rounds,
        time,
        max_players,
        custom_words,
        custom_words_probability,
        public_game,
        vote_kick_enabled,
        players,
        started,
        creator_id
      };
    case UPDATE_GAME_STATE:
      return { ...state, ...action.payload };

    case UPDATE_GAME_PLAYERS:
      return { ...state, players: action.payload };

    case ADMIN_UPDATED:
      return { ...state, creator_id: action.payload };

    case REMOVE_PLAYER:
      return { ...state, players: state.players.filter(player => player.id !== action.payload) };

    default:
      return state;
  }
};

export default gameReducer;

File: lib/pictionary/game.ex
defmodule Pictionary.Game do
  defstruct([
    id: Ecto.UUID.generate(),
    rounds: 3,
    time: 60,
    max_players: 10,
    custom_words: [],
    custom_words_probability: 50,
    public_game: true,
    vote_kick_enabled: true,
    players: MapSet.new(),
    started: false,
    creator_id: nil,
    created_at: DateTime.utc_now(),
    updated_at: DateTime.utc_now(),
  ])

  @type t :: %__MODULE__{
          id: String.t(),
          rounds: number(),
          time: number(),
          max_players: number(),
          custom_words: list(),
          custom_words_probability: number(),
          public_game: boolean(),
          vote_kick_enabled: boolean(),
          players: MapSet.t(),
          creator_id: String.t(),
          created_at: t(),
          updated_at: t(),
          started: boolean()
        }
end
